Data/HashMap/Internal.hs

```diff
@@ -1127,13 +1124,13 @@ delete k m = delete' (hash k) k m
 {-# INLINABLE delete #-}
 
 delete' :: Eq k => Hash -> k -> HashMap k v -> HashMap k v
-delete' h0 k0 m0 = go h0 k0 0 m0
+delete' h0 !k0 m0 = go h0 k0 0 m0
   where
-    go !_ !_ !_ Empty = Empty
+    go _ _ _ Empty = Empty
     go h k _ t@(Leaf hy (L ky _))
         | hy == h && ky == k = Empty
         | otherwise          = t
-    go h k s t@(BitmapIndexed b ary)
+    go !h !k !s t@(BitmapIndexed b ary)
         | b .&. m == 0 = t
         | otherwise =
             let !st = A.index ary i
```

Comment:
What's the net effect of all these bang pattern changes? Do they change compiled Core? Do they change demand signatures, unfoldings, or unfolding guidance?

Reply:
These do change compiled Core. I just checked `insert` and `delete`. There are positive changes with `insert`. The branche's `insert` loop fuses:

```haskell
Data.HashMap.Internal.$winsert'
  = \ (@ k_sGwF)
      (@ v_sGwG)
      (w_sGwH :: Eq k_sGwF)
      (ww_sGwO :: Exts.Word#)
      (w1_sGwJ :: k_sGwF)
      (w2_sGwK :: v_sGwG)
      (w3_sGwL :: HashMap k_sGwF v_sGwG) ->
      letrec {
        $s$wgo_sP6x [Occ=LoopBreaker]
          :: HashMap k_sGwF v_sGwG
             -> Int# -> v_sGwG -> k_sGwF -> Exts.Word# -> HashMap k_sGwF v_sGwG
        [LclId, Arity=5, Str=<S,1*U><L,U><L,U><L,U><L,U>, Unf=OtherCon []]
        $s$wgo_sP6x
          = \ (sc_sP6w :: HashMap k_sGwF v_sGwG)
              (sc1_sP6v :: Int#)
              (sc2_sP6u :: v_sGwG)
              (sc3_sP6t :: k_sGwF)
              (sc4_sP6s :: Exts.Word#) ->
              case sc_sP6w of wild_X8H {
...
                  }
              }; } in
      $s$wgo_sP6x w3_sGwL 0# w2_sGwK w1_sGwJ ww_sGwO
```

note only 1 `wgo....` defined in the `letrec`, master uses two:

```haskell
Data.HashMap.Internal.$winsert'
  = \ (@ k_sGhe)
      (@ v_sGhf)
      (w_sGhg :: Eq k_sGhe)
      (ww_sGhn :: Exts.Word#)
      (w1_sGhi :: k_sGhe)
      (w2_sGhj :: v_sGhf)
      (w3_sGhk :: HashMap k_sGhe v_sGhf) ->
      letrec {
        $s$wgo_sOLH [Occ=LoopBreaker]
          :: Exts.Word#
             -> Exts.SmallArray# (HashMap k_sGhe v_sGhf)
             -> Int#
             -> v_sGhf
             -> k_sGhe
             -> Exts.Word#
             -> HashMap k_sGhe v_sGhf
        [LclId,
         Arity=6,
         Str=<L,U><L,U><L,U><L,U><S,1*U><L,U>,
         Unf=OtherCon []]
        $s$wgo_sOLH
          = \ (sc_sOLF :: Exts.Word#)
              (sc1_sOLG :: Exts.SmallArray# (HashMap k_sGhe v_sGhf))
              (sc2_sOLE :: Int#)
              (sc3_sOLD :: v_sGhf)
              (sc4_sOLC :: k_sGhe)
              (sc5_sOLB :: Exts.Word#) ->
              case sc4_sOLC of k1_XcAI { __DEFAULT ->
...
        $wgo1_sGhd
          = \ (ww1_sGh7 :: Exts.Word#)
              (w4_sGh1 :: k_sGhe)
              (w5_sGh2 :: v_sGhf)
              (ww2_sGhb :: Int#)
              (w6_sGh4 :: HashMap k_sGhe v_sGhf) ->
              case w4_sGh1 of k1_XcAI { __DEFAULT ->
              case w6_sGh4 of wild_X8X {
...
              }
              }; } in
      $wgo1_sGhd ww_sGhn w1_sGhi w2_sGhj 0# w3_sGhk
```

where `go_sGhd` calls `go_sOLH` in the `Collision` case. This is the case that gets fused on the branch.

For whatever reason the bang on `delete` stops the `worker/wrapper` transformation:
branch:
```haskell
delete'
  = \ (@ k_agXw)
      (@ v_agXx)
      ($dEq_agXz :: Eq k_agXw)
      (h0_acCw :: Hash)
      (k0_acCx :: k_agXw)
      (m0_acCy :: HashMap k_agXw v_agXx) ->
      case k0_acCx of k1_XcG7 { __DEFAULT ->
      letrec {
        $sgo_sOOZ [Occ=LoopBreaker]
          :: HashMap k_agXw v_agXx
             -> Int# -> k_agXw -> Exts.Word# -> HashMap k_agXw v_agXx
        [LclId, Arity=4, Str=<S,1*U><L,U><L,U><L,U>, Unf=OtherCon []]
        $sgo_sOOZ
```

master:
```haskell
Data.HashMap.Internal.$wdelete'
  = \ (@ k_sGya)
      (@ v_sGyb)
      (w_sGyc :: Eq k_sGya)
      (ww_sGyi :: Exts.Word#)
      (w1_sGye :: k_sGya)
      (w2_sGyf :: HashMap k_sGya v_sGyb) ->
      letrec {
        $wgo1_sGy9 [InlPrag=NOUSERINLINE[2], Occ=LoopBreaker]
          :: Exts.Word#
             -> k_sGya -> Int# -> HashMap k_sGya v_sGyb -> HashMap k_sGya v_sGyb
        [LclId, Arity=4, Str=<L,U><S,1*U><L,U><S,1*U>, Unf=OtherCon []]
        $wgo1_sGy9
```

Notice also that on `master` the `Hash` input gets unboxed due to the WW transformation.

I think that making sure `delete` gets Worker/Wrapper'd before merging is a good idea. I'm also fine with just reverting the bangs and opening another PR for any more changes.